File: backend/app/comps.py
"""Comparable-listing ranking for the property detail page.

Pure functions only — no I/O, no Realtor calls. Given a subject property and a
list of candidate for-sale listings (from the per-ZIP area cache), gate to
strict appraisal-style comparables and rank the survivors by a weighted
similarity score. The endpoint layer feeds in the cached listings; this module
just decides which are comparable and in what order.
"""
from __future__ import annotations

import math
from typing import Any

# Dissimilarity weights. The score normalizes by the weights actually used, so a
# candidate missing a dimension isn't unfairly penalized — that dimension just
# drops out of both numerator and denominator.
WEIGHTS = {
    "sqft": 0.35,
    "distance": 0.15,
    "year_built": 0.15,
    "beds": 0.15,
    "baths": 0.10,
    "lot_sqft": 0.10,
}

# Difference that maps to a fully-dissimilar (1.0) contribution per dimension.
SQFT_FULL_DIFF = 1.0      # relative: 100% larger/smaller
LOT_FULL_DIFF = 1.0       # relative
YEAR_FULL_DIFF = 50.0     # 50 years apart
BEDS_FULL_DIFF = 3.0
BATHS_FULL_DIFF = 3.0
DISTANCE_FULL_DIFF = 5.0  # miles (a ZIP is usually well under this)

# Strict appraisal-style gates.
SQFT_GATE = 0.25          # ±25% living area
SQFT_GATE_RELAXED = 0.40  # fallback widen
BEDS_GATE = 1             # ±1 bedroom

_NEAREST_LABEL = "showing the nearest matches"


def _f(v: Any) -> float | None:
    try:
        return None if v is None else float(v)
    except (TypeError, ValueError):
        return None


def _clamp01(x: float) -> float:
    return 0.0 if x < 0 else 1.0 if x > 1 else x


def haversine_miles(lat1: Any, lon1: Any, lat2: Any, lon2: Any) -> float | None:
    """Great-circle distance in miles, or None if any coordinate is missing."""
    coords = [_f(lat1), _f(lon1), _f(lat2), _f(lon2)]
    if any(c is None for c in coords):
        return None
    lat1, lon1, lat2, lon2 = coords
    r = 3958.7613  # earth radius, miles
    p1, p2 = math.radians(lat1), math.radians(lat2)
    dphi = math.radians(lat2 - lat1)
    dlmb = math.radians(lon2 - lon1)
    h = math.sin(dphi / 2) ** 2 + math.cos(p1) * math.cos(p2) * math.sin(dlmb / 2) ** 2
    return 2 * r * math.asin(min(1.0, math.sqrt(h)))


def price_per_sqft(price: Any, sqft: Any) -> int | None:
    p, s = _f(price), _f(sqft)
    if p is None or not s:
        return None
    return int(round(p / s))


def _match_pct(subject: dict, c: dict) -> tuple[int, float | None]:
    """Return (match %, distance_mi). Only dimensions present on both sides count."""
    dims: list[tuple[float, float]] = []  # (weight, normalized 0..1 difference)

    def add(weight: float, sval: Any, cval: Any, full: float, relative: bool = False) -> None:
        sv, cv = _f(sval), _f(cval)
        if sv is None or cv is None:
            return
        if relative:
            if not sv:
                return
            diff = abs(cv - sv) / sv
        else:
            diff = abs(cv - sv)
        dims.append((weight, _clamp01(diff / full)))

    add(WEIGHTS["sqft"], subject.get("sqft"), c.get("sqft"), SQFT_FULL_DIFF, relative=True)
    add(WEIGHTS["year_built"], subject.get("year_built"), c.get("year_built"), YEAR_FULL_DIFF)
    add(WEIGHTS["beds"], subject.get("beds"), c.get("beds"), BEDS_FULL_DIFF)
    add(WEIGHTS["baths"], subject.get("baths"), c.get("baths"), BATHS_FULL_DIFF)
    add(WEIGHTS["lot_sqft"], subject.get("lot_sqft"), c.get("lot_sqft"), LOT_FULL_DIFF, relative=True)

    dist = haversine_miles(
        subject.get("latitude"), subject.get("longitude"),
        c.get("latitude"), c.get("longitude"),
    )
    if dist is not None:
        dims.append((WEIGHTS["distance"], _clamp01(dist / DISTANCE_FULL_DIFF)))

    if not dims:
        return 0, dist
    total_w = sum(w for w, _ in dims)
    dissimilarity = sum(w * d for w, d in dims) / total_w
    return round(100 * (1 - dissimilarity)), dist


def _passes_user_filter(c: dict, f: dict) -> bool:
    """User-selected filter pills (min/max price + sqft, min beds + baths).

    Applied *before* the appraisal gate ladder so ranking draws the best comps
    that match the filter from the whole cached pool. Mirrors the frontend rule:
    a candidate missing a dimension is never excluded on that dimension — only
    listings that actively fall outside a band the user set are dropped.
    """
    lp = _f(c.get("list_price"))
    if lp is not None:
        if f.get("min_price") is not None and lp < f["min_price"]:
            return False
        if f.get("max_price") is not None and lp > f["max_price"]:
            return False
    cs = _f(c.get("sqft"))
    if cs is not None:
        if f.get("min_sqft") is not None and cs < f["min_sqft"]:
            return False
        if f.get("max_sqft") is not None and cs > f["max_sqft"]:
            return False
    cb = _f(c.get("beds"))
    if cb is not None and f.get("min_beds") is not None and cb < f["min_beds"]:
        return False
    cba = _f(c.get("baths"))
    if cba is not None and f.get("min_baths") is not None and cba < f["min_baths"]:
        return False
    return True


def _passes_gates(subject: dict, c: dict, sqft_tol: float | None, beds_gate: bool) -> bool:
    st, ct = subject.get("property_type"), c.get("property_type")
    if st and ct and st != ct:
        return False
    s_sqft, c_sqft = _f(subject.get("sqft")), _f(c.get("sqft"))
    if c_sqft is None:  # candidate must have living area to be a comp
        return False
    if sqft_tol is not None and s_sqft:
        if abs(c_sqft - s_sqft) / s_sqft > sqft_tol:
            return False
    if beds_gate:
        s_beds, c_beds = _f(subject.get("beds")), _f(c.get("beds"))
        if s_beds is not None and c_beds is not None and abs(c_beds - s_beds) > BEDS_GATE:
            return False
    return True


def rank_comparables(subject: dict | None, listings: list[dict] | None,
                     filters: dict | None = None, limit: int | None = 6) -> dict:
    """Gate + rank candidate listings into strict appraisal-style comparables.

    Returns ``{comps, relaxed, limited, subject_price_per_sqft}``:
      - ``comps``: ranked list (best match first), each candidate dict copied
        with ``comp_score`` (match %), ``distance_mi``, and ``price_per_sqft``.
        Capped to ``limit`` (``None`` returns the whole ranked set).
      - ``relaxed``: ``None`` when the strict gates produced comps, else a short
        human label naming which fallback rung was used.
      - ``limited``: True when the subject itself lacks ``sqft`` (size gating and
        the $/sqft reference can't be computed).
      - ``subject_price_per_sqft``: the subject's own $/sqft reference line.

    ``filters`` (optional user-selected pills) pre-filters the candidate pool via
    ``_passes_user_filter`` *before* the gate ladder, so the strictest appraisal
    rung is computed over only the listings the user asked for — the returned
    comps are the best matches that are BOTH appraisal-comparable and in-filter.

    Fallback ladder: keep the strictest rung that yields ANY comp, so a thin ZIP
    shows a few precise comps rather than diluting with loose ones; only drop to
    the next rung when the current one is empty.
    """
    listings = [l for l in (listings or []) if isinstance(l, dict)]
    if filters:
        listings = [c for c in listings if _passes_user_filter(c, filters)]
    subject = subject or {}

    # (label, sqft tolerance, enforce beds gate). label None == strict, no note.
    ladder = [
        (None, SQFT_GATE, True),
        ("widened the size range to ±40%", SQFT_GATE_RELAXED, True),
        ("dropped the bedroom limit", SQFT_GATE_RELAXED, False),
        (_NEAREST_LABEL, None, False),
    ]

    chosen: list[dict] = []
    relaxed: str | None = None
    for label, sqft_tol, beds_gate in ladder:
        is_last = label == _NEAREST_LABEL
        passed = listings if is_last else [
            c for c in listings if _passes_gates(subject, c, sqft_tol, beds_gate)
        ]
        if passed or is_last:
            chosen = passed
            relaxed = label
            break

    scored: list[dict] = []
    for c in chosen:
        match_pct, dist = _match_pct(subject, c)
        out = dict(c)
        out["comp_score"] = match_pct
        out["distance_mi"] = round(dist, 2) if dist is not None else None
        out["price_per_sqft"] = price_per_sqft(c.get("list_price"), c.get("sqft"))
        scored.append(out)

    scored.sort(key=lambda x: (
        -x["comp_score"],
        x["distance_mi"] if x["distance_mi"] is not None else 9e9,
        x.get("list_price") or 0,
    ))

    subject_price = subject.get("list_price")
    if subject_price is None:
        subject_price = subject.get("best_current_estimate")

    return {
        "comps": scored if limit is None else scored[:limit],
        "relaxed": relaxed,
        "limited": _f(subject.get("sqft")) is None,
        "subject_price_per_sqft": price_per_sqft(subject_price, subject.get("sqft")),
    }


def comp_domain(subject: dict | None, listings: list[dict] | None) -> dict:
    """Stable filter-slider domain for the comp module: the price/sqft spread of
    the *unfiltered* comp set (the full ranked pool, not just the shown page).

    Returned to the frontend so the Price/Sqft sliders span the real comp
    universe and don't rescale every time the user applies a filter (which the
    server now honors by re-ranking). ``count`` lets the UI tell "this ZIP has no
    comps" apart from "your filters excluded them all".
    """
    base = rank_comparables(subject, listings, limit=None)["comps"]
    prices = [int(round(v)) for v in (_f(c.get("list_price")) for c in base) if v is not None]
    sqfts = [int(round(v)) for v in (_f(c.get("sqft")) for c in base) if v is not None]
    return {"prices": prices, "sqfts": sqfts, "count": len(base)}

File: src/server/comps.ts
// Comparable-listing ranking for the property detail page — ported from
// backend/app/comps.py.
//
// Pure functions only — no I/O, no Realtor calls. Gate candidate for-sale
// listings to strict appraisal-style comparables and rank survivors by a
// weighted similarity score (AGENTS.md rule #15). Keep the scoring here — one
// pure module; don't fork it.

import { pyround, pyround2 } from './pyround'

// Dissimilarity weights. The score normalizes by the weights actually used, so a
// candidate missing a dimension isn't unfairly penalized.
const WEIGHTS: Record<string, number> = {
  sqft: 0.35,
  distance: 0.15,
  year_built: 0.15,
  beds: 0.15,
  baths: 0.1,
  lot_sqft: 0.1,
}

// Difference that maps to a fully-dissimilar (1.0) contribution per dimension.
const SQFT_FULL_DIFF = 1.0 // relative: 100% larger/smaller
const LOT_FULL_DIFF = 1.0 // relative
const YEAR_FULL_DIFF = 50.0 // 50 years apart
const BEDS_FULL_DIFF = 3.0
const BATHS_FULL_DIFF = 3.0
const DISTANCE_FULL_DIFF = 5.0 // miles

// Strict appraisal-style gates.
const SQFT_GATE = 0.25 // ±25% living area
const SQFT_GATE_RELAXED = 0.4 // fallback widen
const BEDS_GATE = 1 // ±1 bedroom

const _NEAREST_LABEL = 'showing the nearest matches'

function _f(v: any): number | null {
  if (v === null || v === undefined) return null
  const n = Number(v)
  return Number.isFinite(n) ? n : null
}

function _clamp01(x: number): number {
  return x < 0 ? 0.0 : x > 1 ? 1.0 : x
}

function radians(deg: number): number {
  return (deg * Math.PI) / 180
}

// Great-circle distance in miles, or null if any coordinate is missing.
export function haversine_miles(lat1: any, lon1: any, lat2: any, lon2: any): number | null {
  const coords = [_f(lat1), _f(lon1), _f(lat2), _f(lon2)]
  if (coords.some((c) => c === null)) return null
  const [la1, lo1, la2, lo2] = coords as number[]
  const r = 3958.7613 // earth radius, miles
  const p1 = radians(la1)
  const p2 = radians(la2)
  const dphi = radians(la2 - la1)
  const dlmb = radians(lo2 - lo1)
  const h =
    Math.sin(dphi / 2) ** 2 + Math.cos(p1) * Math.cos(p2) * Math.sin(dlmb / 2) ** 2
  return 2 * r * Math.asin(Math.min(1.0, Math.sqrt(h)))
}

export function price_per_sqft(price: any, sqft: any): number | null {
  const p = _f(price)
  const s = _f(sqft)
  if (p === null || !s) return null
  return Math.trunc(pyround(p / s))
}

// Return [match %, distance_mi]. Only dimensions present on both sides count.
function _match_pct(subject: any, c: any): [number, number | null] {
  const dims: [number, number][] = [] // (weight, normalized 0..1 difference)

  const add = (weight: number, sval: any, cval: any, full: number, relative = false): void => {
    const sv = _f(sval)
    const cv = _f(cval)
    if (sv === null || cv === null) return
    let diff: number
    if (relative) {
      if (!sv) return
      diff = Math.abs(cv - sv) / sv
    } else {
      diff = Math.abs(cv - sv)
    }
    dims.push([weight, _clamp01(diff / full)])
  }

  add(WEIGHTS.sqft, subject.sqft, c.sqft, SQFT_FULL_DIFF, true)
  add(WEIGHTS.year_built, subject.year_built, c.year_built, YEAR_FULL_DIFF)
  add(WEIGHTS.beds, subject.beds, c.beds, BEDS_FULL_DIFF)
  add(WEIGHTS.baths, subject.baths, c.baths, BATHS_FULL_DIFF)
  add(WEIGHTS.lot_sqft, subject.lot_sqft, c.lot_sqft, LOT_FULL_DIFF, true)

  const dist = haversine_miles(subject.latitude, subject.longitude, c.latitude, c.longitude)
  if (dist !== null) {
    dims.push([WEIGHTS.distance, _clamp01(dist / DISTANCE_FULL_DIFF)])
  }

  if (!dims.length) return [0, dist]
  const total_w = dims.reduce((acc, [w]) => acc + w, 0)
  const dissimilarity = dims.reduce((acc, [w, d]) => acc + w * d, 0) / total_w
  return [pyround(100 * (1 - dissimilarity)), dist]
}

// User-selected filter pills (min/max price + sqft, min beds + baths). Applied
// BEFORE the appraisal gate ladder. A candidate missing a dimension is never
// excluded on that dimension.
function _passes_user_filter(c: any, f: any): boolean {
  const lp = _f(c.list_price)
  if (lp !== null) {
    if (f.min_price !== null && f.min_price !== undefined && lp < f.min_price) return false
    if (f.max_price !== null && f.max_price !== undefined && lp > f.max_price) return false
  }
  const cs = _f(c.sqft)
  if (cs !== null) {
    if (f.min_sqft !== null && f.min_sqft !== undefined && cs < f.min_sqft) return false
    if (f.max_sqft !== null && f.max_sqft !== undefined && cs > f.max_sqft) return false
  }
  const cb = _f(c.beds)
  if (cb !== null && f.min_beds !== null && f.min_beds !== undefined && cb < f.min_beds) return false
  const cba = _f(c.baths)
  if (cba !== null && f.min_baths !== null && f.min_baths !== undefined && cba < f.min_baths) return false
  return true
}

function _passes_gates(subject: any, c: any, sqft_tol: number | null, beds_gate: boolean): boolean {
  const st = subject.property_type
  const ct = c.property_type
  if (st && ct && st !== ct) return false
  const s_sqft = _f(subject.sqft)
  const c_sqft = _f(c.sqft)
  if (c_sqft === null) return false // candidate must have living area to be a comp
  if (sqft_tol !== null && s_sqft) {
    if (Math.abs(c_sqft - s_sqft) / s_sqft > sqft_tol) return false
  }
  if (beds_gate) {
    const s_beds = _f(subject.beds)
    const c_beds = _f(c.beds)
    if (s_beds !== null && c_beds !== null && Math.abs(c_beds - s_beds) > BEDS_GATE) return false
  }
  return true
}

// Gate + rank candidate listings into strict appraisal-style comparables.
// Returns {comps, relaxed, limited, subject_price_per_sqft}.
export function rank_comparables(
  subject: any | null,
  listings: any[] | null,
  filters: any | null = null,
  limit: number | null = 6,
): any {
  let list = (listings || []).filter((l) => l !== null && typeof l === 'object' && !Array.isArray(l))
  if (filters) {
    list = list.filter((c) => _passes_user_filter(c, filters))
  }
  subject = subject || {}

  // [label, sqft tolerance, enforce beds gate]. label null == strict, no note.
  const ladder: [string | null, number | null, boolean][] = [
    [null, SQFT_GATE, true],
    ['widened the size range to ±40%', SQFT_GATE_RELAXED, true],
    ['dropped the bedroom limit', SQFT_GATE_RELAXED, false],
    [_NEAREST_LABEL, null, false],
  ]

  let chosen: any[] = []
  let relaxed: string | null = null
  for (const [label, sqft_tol, beds_gate] of ladder) {
    const is_last = label === _NEAREST_LABEL
    const passed = is_last ? list : list.filter((c) => _passes_gates(subject, c, sqft_tol, beds_gate))
    if (passed.length || is_last) {
      chosen = passed
      relaxed = label
      break
    }
  }

  const scored: any[] = []
  for (const c of chosen) {
    const [match_pct, dist] = _match_pct(subject, c)
    const out = { ...c }
    out.comp_score = match_pct
    out.distance_mi = dist !== null ? pyround2(dist) : null
    out.price_per_sqft = price_per_sqft(c.list_price, c.sqft)
    scored.push(out)
  }

  scored.sort((a, b) => {
    if (a.comp_score !== b.comp_score) return b.comp_score - a.comp_score
    const da = a.distance_mi !== null && a.distance_mi !== undefined ? a.distance_mi : 9e9
    const db = b.distance_mi !== null && b.distance_mi !== undefined ? b.distance_mi : 9e9
    if (da !== db) return da - db
    return (a.list_price || 0) - (b.list_price || 0)
  })

  let subject_price = subject.list_price
  if (subject_price === null || subject_price === undefined) {
    subject_price = subject.best_current_estimate
  }

  return {
    comps: limit === null ? scored : scored.slice(0, limit),
    relaxed,
    limited: _f(subject.sqft) === null,
    subject_price_per_sqft: price_per_sqft(subject_price, subject.sqft),
  }
}

// Stable filter-slider domain for the comp module: the price/sqft spread of the
// UNFILTERED comp set (the full ranked pool, not just the shown page).
export function comp_domain(subject: any | null, listings: any[] | null): any {
  const base = rank_comparables(subject, listings, null, null).comps
  const prices = base
    .map((c: any) => _f(c.list_price))
    .filter((v: number | null) => v !== null)
    .map((v: number) => Math.trunc(pyround(v)))
  const sqfts = base
    .map((c: any) => _f(c.sqft))
    .filter((v: number | null) => v !== null)
    .map((v: number) => Math.trunc(pyround(v)))
  return { prices, sqfts, count: base.length }
}
